File: content/snippets/python/design_patterns/prototype.py
from dataclasses import dataclass
from copy import deepcopy


@dataclass
class Report:
    title: str
    tags: list[str]

    def clone(self) -> "Report":
        return deepcopy(self)


template = Report("weekly", ["ops", "summary"])
copy = template.clone()

print(template)
print(copy)

File: content/snippets/typescript/design_patterns/prototype.ts
interface Prototype<T> {
  clone(): T;
}

class Report implements Prototype<Report> {
  constructor(
    public readonly title: string,
    public readonly tags: string[],
  ) {}

  clone(): Report {
    return new Report(this.title, [...this.tags]);
  }
}

const template = new Report('weekly', ['ops', 'summary']);
const copy = template.clone();

console.log(template);
console.log(copy);